neps/nep-0413.md

```diff
@@ -0,0 +1,176 @@
+  recipient: string; // The same recipient passed in `SignMessageParams.recipient`
+  callbackUrl?: string // The same callbackUrl passed in `SignMessageParams.callbackUrl`
+}
+```
+
+### Signature
+In order to create a signature, `signMessage` must:
+1. Create a `Payload` object. 
+2. Convert the `payload` into its [Borsh Representation](https://borsh.io).
+3. Prepend the 4-bytes borsh representation of $2^{31}+413$, as the [prefix tag](https://github.com/near/NEPs/pull/461).
+4. Compute the `SHA256` hash of the serialized-prefix + serialized-tag.
+5. Sign the resulting `SHA256` hash from step 3 using a **full-access** key.
+
+> If the wallet does not hold any `full-access` keys, then it must return an error.
+
+### Example
+Assuming that the `signMessage` method was invoked, and that:
+- The input `message` is `"hi"`
+- The input `nonce` is `[0,...,31]`
+- The input `recipient` is `"myapp.com"`
+- The callbackUrl is `"myapp.com/callback"`
+- The wallet stores a full-access private key
+
+The wallet must construct and sign the following `SHA256` hash:
+
+```jsx
+// 2**31 + 413 == 2147484061
+sha256.hash(Borsh.serialize<u32>(2147484061) + Borsh.serialize(Payload{message:"hi", nonce:[0,...,31], recipient:"myapp.com", callbackUrl: "myapp.com/callback"}))
+```
+
+### Output Interface
+`signMessage` must return an object containing the **base64** representation of the `signature`, and all the data necessary to verify such signature. 
+
+```jsx
+interface SignedMessage {
+  accountId: string; // The account name to which the publicKey corresponds as plain text (e.g. "alice.near")
+  publicKey: string; // The public counterpart of the key used to sign, expressed as a string with format "<key-type>:<base58-key-bytes>" (e.g. "ed25519:6TupyNrcHGTt5XRLmHTc2KGaiSbjhQi1KHtCXTgbcr4Y") 
+  signature: string; // The base64 representation of the signature.
+  state?: string; // Optional, applicable to browser wallets (e.g. MyNearWallet). The same state passed in SignMessageParams.
+}
+```
+
+### Returning the signature
+#### Web Wallets
+Web Wallets, such as [MyNearWallet](https://mynearwallet.com), should directly return the `SignedMessage` to the `SignMessageParams.callbackUrl`, passing the `accountId`,`publicKey`, `signature` and the state as URL fragments. This is: `<callbackUrl>#accountId=<accountId>&publicKey=<publicKey>&signature=<signature>&state=<state>`.
+
+If the signing process fails, then the wallet must return an error message and the state as string fragments: `<callbackUrl>#error=<error-message-string>&state=<state>`.
+
+#### Other Wallets
+Non-web Wallets, such as [Ledger](https://www.ledger.com) can directly return the `SignedMessage` (in preference as a JSON object) and raise an error on failure.
+
+## References
+A full example on how to implement the `signMessage` method can be [found here](https://github.com/gagdiez/near-login/blob/main/tests/authentication/auth.ava.ts#L27-#L65).
```

Comment:
@gagdiez is there any way/example on how to verify the signature for browser wallets?

The [verifySignature ](https://github.com/gagdiez/near-login/blob/main/authenticate/wallet-authenticate.js#L19) function in the example needs the `message`, `nonce`, `recipient` and `callbackUrl` to re-construct the payload which was actually signed but we lose all this data on redirect?

What we get back in the URL is only the `SignedMessage`.

Reply:
with the current approach, you will need the user to pass it back to you. If you are implementing an app in which the user is asked to sign something, then you will need to persist somewhere the parameters, redirect the user to the wallet, get the signature, and relay it alongside the rest of the params